### PR TITLE
Properly respect externally managed annotation

### PR DESCRIPTION
**Pull request by alvaroaleman**
**What type of PR is this?**

/kind bug

**What this PR does / why we need it**:

We already have a predicate for events from AzureClusters that skip
their reconciliation if they are managed externally. We do however not
have such a check for events that come from a Cluster, resulting in
us doing a reconciliation for AzureClusters anyways.
    
The approach is copied from capa:
https://github.com/kubernetes-sigs/cluster-api-provider-aws/blob/cf6fea721385d490b570b4cf119defea705b3320/controllers/awscluster_controller.go#L331-L377

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:

_Please confirm that if this PR changes any image versions, then that's the sole change this PR makes._

**TODOs**:

- [ ] squashed commits
- [ ] includes documentation
- [ ] adds unit tests

**Release note**:

```release-note
The azurecluster reconciler now always respects the ExternallyManaged annotation
```

### CODE REVIEW

File: controllers/azurecluster_controller.go
Comment (sonasingh46) [inline controllers/azurecluster_controller.go:102]: Any specific reason to pass `log` as an arg? I think you can get an instance of logger from ctx itself.
Ref: https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/main/controllers/azurecluster_controller.go#L76

File: controllers/azurecluster_controller.go
Comment (sonasingh46) [inline controllers/azurecluster_controller.go:327]: nit: AWSCluster --> AzureCluster

**Comment by enxebre** [comment]
Can we please make sure the tag can't be removed via webhook? see https://github.com/kubernetes-sigs/cluster-api-provider-aws/pull/2383#diff-bede819718d1d34b23f65d8e8aad84aae8aa8311be130d14c8dc81d558ec44e1R108

File: controllers/azurecluster_controller.go
Comment (enxebre) [inline controllers/azurecluster_controller.go:102]: wouldn't we need to just add an event filter as in https://github.com/kubernetes-sigs/cluster-api-provider-azure/pull/2023/files#diff-c9f392795aea4d131d38cf4a128c8f6be728ca2ec8b8307cad6ecf1fb7ca5da1L91?

**Comment by enxebre** [comment]
For ref this was already documented here https://github.com/kubernetes-sigs/cluster-api-provider-azure/pull/1807

File: controllers/azurecluster_controller.go
Comment (alvaroaleman) [reply controllers/azurecluster_controller.go]: No specific reason, updated, lmk what you think

File: controllers/azurecluster_controller.go
Comment (alvaroaleman) [reply controllers/azurecluster_controller.go]: Fixed

File: controllers/azurecluster_controller.go
Comment (alvaroaleman) [reply controllers/azurecluster_controller.go]: The reason a predicate can not be used is that the the annotation only exists on the azurecluster, but not on the capi cluster, so we have to fetch the corresponding azure cluster in order to find out if it has the annotation. That predicate only works for a watch on the azure cluster

**Comment by alvaroaleman** [comment]
@sonasingh46 @enxebre addressed the feedback, please take another look

File: controllers/azurecluster_controller.go
Comment (CecileRobertMichon) [inline controllers/azurecluster_controller.go:302]: in CAPZ we have a logger used for tracing, see #1806 

```suggestion
	ctx, log, done := tele.StartSpanWithLogger(ctx, "controllers.AzureClusterReconciler.clusterToInfrastructureMapFunc")
	defer done()
```

File: controllers/azurecluster_controller.go
Comment (CecileRobertMichon) [inline controllers/azurecluster_controller.go:303]: this function is in the azurecluster_controller file, if it's meant to be used by multiple controllers should it be somewhere common instead? maybe helpers.go?

Also side question, if this needed for other providers too (sounds like it was added to CAPA), why not change the `ClusterToInfrastructureMapFunc` function in CAPI directly to skip externally managed clusters instead of overriding it with our own `clusterToInfrastructureMapFunc` func?

Comment (alvaroaleman) [reply controllers/azurecluster_controller.go:303]: Good point, opened https://github.com/kubernetes-sigs/cluster-api/pull/6039 for that and will update this and capa once it merges

Comment (alvaroaleman) [reply controllers/azurecluster_controller.go]: Updated to bump cluster-api and use the new ClusterToInfrastructureMapFuncWithExternallyManagedCheck, please have another look.

**Review by CecileRobertMichon — COMMENTED** [review]
lgtm

should we wait for an actual release of CAPI to merge this?

**Comment by alvaroaleman** [comment]
> should we wait for an actual release of CAPI to merge this?

Not sure, what is the common approach for a situation like this where we need a bugfix from an unreleased version?

**Review by sonasingh46 — APPROVED** [review]
lgtm

**Comment by CecileRobertMichon** [comment]
/hold

let's wait for a CAPI release since we don't want to release CAPZ with an unreleased version of CAPI

File: go.mod
Comment (CecileRobertMichon) [inline go.mod:45]: @alvaroaleman v1.1.1 is out with this change, we should be able to update and merge this now

thanks again

Comment (alvaroaleman) [reply go.mod]: Thanks, updated

**Review by CecileRobertMichon — COMMENTED** [review]
/lgtm
/approve
/hold cancel

**Pull request merged**
